src/sentry/integrations/utils/metrics.py

```diff
@@ -0,0 +1,147 @@
+from abc import ABC, abstractmethod
+from collections.abc import Mapping
+from enum import Enum
+from types import TracebackType
+from typing import Any, Self
+
+from django.conf import settings
+
+from sentry.utils import metrics
+
+
+class EventLifecycleOutcome(Enum):
+    STARTED = "STARTED"
+    HALTED = "HALTED"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+    def __str__(self) -> str:
+        return self.value.lower()
+
+
+class EventLifecycleMetric(ABC):
+    """Information about an event to be measured.
+
+    This class is intended to be used across different integrations that share the
+    same business concern. Generally a subclass would represent one business concern
+    (such as MessagingInteractionEvent, which extends this class and is used in the
+    `slack`, `msteams`, and `discord` integration packages).
+    """
+
+    @abstractmethod
+    def get_key(self, outcome: EventLifecycleOutcome) -> str:
+        """Construct the metrics key that will represent this event."""
+        raise NotImplementedError
+
+    def get_extras(self) -> Mapping[str, Any]:
```

Comment:
would we pass this into `EventLifecycle` to emit with a logger at success / failure points? just curious about how this might help with debugging!

Reply:
Good point! I held off because I wasn't sure exactly when/how we'd want to log that information, but here's a first pass: https://github.com/getsentry/sentry/pull/78482/commits/9ae594772bfd55873c1d2a60fe73785bd3a008dc